capsules/src/sdcard.rs

```diff
@@ -0,0 +1,1424 @@
+//! Provide capsule driver for accessing an SD Card.
+//! This allows initialization and block reads or writes on top of SPI
+
+// Resources for SD Card API
+// elm-chan.org/docs/mmc/mmc_e.html
+// alumni.cs.ucr.edu/~amitra/sdcard/Additional/sdcard_appnote_foust.pdf
+// luckyresistor.me/cat-protector/software/sdcard-2/
+// http://users.ece.utexas.edu/~valvano/EE345M/SD_Physical_Layer_Spec.pdf
+
+use core::cell::Cell;
+use core::cmp;
+use kernel::{AppId, AppSlice, Callback, Driver, ReturnCode, Shared};
+use kernel::common::take_cell::{MapCell, TakeCell};
+
+use kernel::hil;
+use kernel::hil::time::Frequency;
+
+// Buffers used for SD card transactions
+// Constraints:
+//  * RXBUFFER must be greater than or equal to TXBUFFER in length
+//  * Both RXBUFFER and TXBUFFER must be longer  than the SD card's block size
+pub static mut TXBUFFER: [u8; 515] = [0; 515];
+pub static mut RXBUFFER: [u8; 515] = [0; 515];
+
+// Buffer for SD card driver
+pub static mut KERNEL_BUFFER: [u8; 512] = [0; 512];
+
+#[allow(dead_code)]
+#[derive(Clone,Copy,Debug,PartialEq)]
+enum SDCmd {
+    CMD0 = 0, // Reset
+    CMD1 = 1, // Generic init
+    CMD8 = 8, // Check voltage range
+    CMD9 = 9, // Read chip specific data (CSD) register
+    CMD12 = 12, // Stop multiple block read
+    CMD16 = 16, // Set blocksize
+    CMD17 = 17, // Read single block
+    CMD18 = 18, // Read multiple blocks
+    CMD24 = 24, // Write single block
+    CMD25 = 25, // Write multiple blocks
+    CMD55 = 55, // Next command will be application specific
+    CMD58 = 58, // Read operation condition register (OCR)
+    ACMD41 = 0x80 + 41, // App-specific Init
```

Comment:
Unfortunately, App-Specific Commands is what the specification named manufacturer specific commands, but then there are some manufacturer specific commands that everybody implements by default these days, hence we are using them to tell card sizes apart. : (